src/codegen/codegen_c_visitor.cpp

```diff
@@ -583,26 +583,34 @@ std::vector<std::string> CodegenCVisitor::ion_read_statements(BlockType type) {
         return ion_read_statements_optimized(type);
     }
     std::vector<std::string> statements;
+    std::unordered_set<std::string> no_doubles;  // used to keep track of the variables to not
+                                                 // have doubles between read/write. Same name
+                                                 // variables are allowed as required in issue
+                                                 // #215
     for (const auto& ion: info.ions) {
         auto name = ion.name;
         for (const auto& var: ion.reads) {
-            if (type == BlockType::Ode && ion.is_ionic_conc(var) && state_variable(var)) {
+            if (type == BlockType::Ode && ion.is_ionic_conc(var) && state_variable(var) ||
+                no_doubles.find(var) != no_doubles.end()) {
                 continue;
             }
             auto variable_names = read_ion_variable_name(var);
             auto first = get_variable_name(variable_names.first);
             auto second = get_variable_name(variable_names.second);
             statements.push_back("{} = {};"_format(first, second));
+            no_doubles.emplace(var);
```

Comment:
```suggestion
            no_doubles.insert(var);
```
You don't need to construct.

Reply:
I do not understand. Both insert and emplace construct a copy of the object. I do not see any advantage with insert over emplace

Reply:
done